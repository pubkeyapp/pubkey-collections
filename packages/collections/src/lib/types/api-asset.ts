export interface ApiAsset {
  id: string
  content: {
    metadata: {
      name: string
      description: string
      attributes: Array<{
        trait_type: string
        value: string
      }>
    }
    files: {
      uri: string
      cdn_uri: string
      mime: string
    }
  }
  grouping: Array<{
    group_type: string
    group_value: string
  }>
}
