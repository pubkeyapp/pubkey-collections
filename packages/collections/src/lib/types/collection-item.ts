import { CollectionTrait } from './collection-trait'

export interface CollectionItem {
  id: string
  name: string
  description?: string
  image?: string
  collectionId: string
  traits: CollectionTrait[]
}
