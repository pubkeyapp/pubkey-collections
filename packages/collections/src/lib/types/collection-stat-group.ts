import { CollectionStat } from './collection-stat'

export interface CollectionStatGroup {
  key: string
  items: CollectionStat[]
}
