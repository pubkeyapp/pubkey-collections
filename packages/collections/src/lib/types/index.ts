export * from './api-asset'
export * from './collection'
export * from './collection-combo'
export * from './collection-item'
export * from './collection-set'
export * from './collection-trait'
export * from './collection-trait-map'
export * from './collection-trait-summary'
