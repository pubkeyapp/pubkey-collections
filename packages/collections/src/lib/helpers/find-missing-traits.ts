import { CollectionTrait, CollectionTraitMap } from '../types'

export function findMissingTraits(collection: CollectionTraitMap, user: CollectionTraitMap): CollectionTrait[] {
  return Object.keys(collection)
    .map((key) => collection[key])
    .reduce(
      (acc, traitValues) => [
        ...acc,
        ...traitValues.filter((value) => !user[value.key]?.find((t) => t.value === value.value)),
      ],
      [],
    )
}
