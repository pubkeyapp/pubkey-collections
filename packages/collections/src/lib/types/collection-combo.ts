import { CollectionItem } from './collection-item'
import { CollectionTrait } from './collection-trait'

export interface CollectionCombo {
  name: string
  group?: boolean
  sortKey?: string
  common?: number
  traits: CollectionTrait[]
  items?: CollectionItem[]
}

export interface CollectionComboGroup {
  name: string
  combos: CollectionCombo[]
}
