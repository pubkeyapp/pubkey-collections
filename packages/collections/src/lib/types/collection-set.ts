import { Collection } from './collection'
import { CollectionItem } from './collection-item'
import { CollectionTraitMap } from './collection-trait-map'

export interface CollectionSet {
  collection: Collection
  items: CollectionItem[]
  traits: CollectionTraitMap
}
