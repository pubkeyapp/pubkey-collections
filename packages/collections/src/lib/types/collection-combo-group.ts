import { CollectionCombo } from './collection-combo'

export interface CollectionComboGroup {
  name: string
  combos: CollectionCombo[]
}
