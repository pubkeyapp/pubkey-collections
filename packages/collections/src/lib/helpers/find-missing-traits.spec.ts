import { CollectionTraitMap } from '../types'
import { findMissingTraits } from './find-missing-traits'

describe('find-missing-traits', () => {
  it('should find missing traits', () => {
    // Arrange
    const traits: CollectionTraitMap = {
      Background: [
        { key: 'Background', value: 'Red' },
        { key: 'Background', value: 'Blue' },
      ],
      Eyes: [{ key: 'Eyes', value: 'Green' }],
      Hair: [],
    }
    const owned: CollectionTraitMap = {
      Background: [{ key: 'Background', value: 'Red' }],
      Eyes: [{ key: 'Eyes', value: 'Green' }],
      Hair: [],
    }

    // Act
    const missing = findMissingTraits(traits, owned)

    // Assert
    expect(missing).toMatchSnapshot()
  })
})
