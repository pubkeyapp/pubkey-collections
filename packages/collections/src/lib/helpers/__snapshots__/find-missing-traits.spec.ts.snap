// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-missing-traits should find missing traits 1`] = `
[
  {
    "key": "Background",
    "value": "Red",
  },
  {
    "key": "Background",
    "value": "Blue",
  },
  {
    "key": "Eyes",
    "value": "Green",
  },
]
`;
